File: backend/java/jx/commons/sochina-commons-base/src/main/java/com/sochina/base/utils/XssUtils.java
package com.sochina.base.utils;

import com.sochina.base.constants.BaseInitResources;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static com.sochina.base.constants.Constants.EMPTY_STRING;

@Component
public class XssUtils {
    private static final Logger LOGGER = LoggerFactory.getLogger(XssUtils.class);

    /**
     * 将字符串中的敏感数据替换成全角字符
     */
    public static String escape(String str) {
        return Optional.ofNullable(str)
                .filter(StringUtils::isNotBlank)
                .map(s -> BaseInitResources.XSS_HALF_MAP.entrySet().stream()
                        .reduce(s, (s1, entry) -> s1.replace(entry.getKey(), entry.getValue()), String::concat))
                .orElse(EMPTY_STRING);
    }

    /**
     * 将字符串中的全角字符还原成半角
     */
    public static String recover(String str) {
        return Optional.ofNullable(str)
                .filter(StringUtils::isNotBlank)
                .map(s -> BaseInitResources.XSS_FULL_MAP.entrySet().stream()
                        .reduce(s, (s1, entry) -> s1.replace(entry.getKey(), entry.getValue()), String::concat))
                .orElse(EMPTY_STRING);
    }

    /**
     * 将字符串中的数据和表达式替换成空格
     */
    public static String clean(String str) {
        if (StringUtils.isBlank(str)) {
            return EMPTY_STRING;
        }
        return BaseInitResources.XSS_PATTERN_MAP.entrySet().stream()
                .reduce(str, (s, entry) -> {
                    Pattern pattern = Pattern.compile(entry.getValue(), Pattern.CASE_INSENSITIVE);
                    Matcher matcher = pattern.matcher(s);
                    return matcher.replaceAll(EMPTY_STRING);
                }, (s1, s2) -> s1);
    }
}

File: backend/kotlin/first/ks-commons-base/src/main/kotlin/com/sochina/base/utils/character/XssUtils.kt
package com.sochina.base.utils.character

import java.util.*
import java.util.regex.Pattern

class XssUtils private constructor(private val commonsProperties: com.sochina.base.properties.CommonsProperties) :
    ICharacterTool {

    companion object {

        @Volatile
        private var instance: XssUtils? = null

        fun getInstance(commonsProperties: com.sochina.base.properties.CommonsProperties): XssUtils {
            return instance ?: synchronized(this) {
                instance ?: XssUtils(commonsProperties).also { instance = it }
            }
        }
    }

    private var XSS_HALF_MAP = mutableMapOf<String, String>()

    private var XSS_FULL_MAP = mutableMapOf<String, String>()

    private var XSS_PATTERN_MAP = mutableMapOf<String, String>()

    private var XSS_SENSITIVE_DATA: List<String>? = emptyList()

    init {
        XSS_PATTERN_MAP = Optional.ofNullable(commonsProperties.xssProperties.pattern).get()
        XSS_SENSITIVE_DATA = commonsProperties.xssProperties.sensitiveData.let { it.customSplit() }
        XSS_SENSITIVE_DATA?.forEach {
            val value = CharUtils.half2Full(it)
            value?.run {
                XSS_HALF_MAP[it] = value
                XSS_FULL_MAP[value] = it
            }
        }
    }

    /**
     * 将字符串中的敏感数据替换成全角字符
     */
    override fun doCheck(str: String): String {
        return str.takeIf { it.isNotBlank() }
            ?.let {
                XSS_HALF_MAP.entries.fold(it) { acc, entry ->
                    acc.replace(entry.key, entry.value)
                }
            }
            ?: ""
    }

    /*
    * 将字符串中的全角字符转换成半角字符
    *  */
    override fun recover(str: String): String {
        return str.takeIf { it.isNotBlank() }
            ?.let {
                XSS_FULL_MAP.entries.fold(it) { acc, entry ->
                    acc.replace(entry.key, entry.value)
                }
            }
            ?: ""
    }

    /*
    * 将所有符合条件的特殊字符转换成空格
    *  */
    fun clean(str: String): String {
        return str.takeIf { it.isNotBlank() }
            ?.let {
                XSS_PATTERN_MAP.entries.fold(it) { acc, entity ->
                    val pattern = Pattern.compile(entity.value, Pattern.CASE_INSENSITIVE)
                    val matcher = pattern.matcher(acc)
                    matcher.replaceAll("")
                }
            }
            ?: ""
    }

    private fun String.customSplit(): List<String> {
        return this.split(",")
    }
}